<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>8. Examples &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API documentation" href="minorg.html" />
    <link rel="prev" title="7. Algorithms" href="algorithms.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">1. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">2. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">3. Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_cli.html">4. Tutorial (Command line)</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_py.html">5. Tutorial (Python)</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">6. Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">7. Algorithms</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">8. Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#example-1">8.1. Example 1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#example-2">8.2. Example 2</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">8. </span>Examples</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="examples">
<h1><span class="section-number">8. </span>Examples<a class="headerlink" href="#examples" title="Permalink to this headline"></a></h1>
<p>In all the following examples, the current directory/working directory is presumed to contain all files in <a class="reference external" href="https://github.com/rlrq/MINORg/tree/master/examples">https://github.com/rlrq/MINORg/tree/master/examples</a>.</p>
<section id="example-1">
<h2><span class="section-number">8.1. </span>Example 1<a class="headerlink" href="#example-1" title="Permalink to this headline"></a></h2>
<p>Generate gRNA in the coding regions of a reference gene.</p>
<ul class="simple">
<li><p>We will use the gene ‘AT1G33560’</p></li>
<li><p>As the default feature type for generating gRNA in is “CDS” (coding region), we do not need to explicitly specify it</p></li>
</ul>
<p><strong>Python code</strong></p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">minorg.MINORg</span> <span class="kn">import</span> <span class="n">MINORg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span> <span class="o">=</span> <span class="n">MINORg</span><span class="p">(</span><span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;example1&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">add_reference</span><span class="p">(</span><span class="s2">&quot;TAIR10&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_ref_TAIR10.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;subset_ref_TAIR10.gff&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="s2">&quot;AT1G33560&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">query_reference</span> <span class="o">=</span> <span class="kc">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">seq</span><span class="p">()</span> <span class="c1">## generate target sequence(s), in this case that of AT1G33560</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">grna</span><span class="p">()</span> <span class="c1">## generate all possible gRNA in target sequence(s)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">filter_feature</span><span class="p">()</span> <span class="c1">## filter gRNA by feature type</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">write_pass_grna_fasta</span><span class="p">()</span> <span class="c1">## write gRNA that pass all filters (in this case, only 1 filter was used)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_minorg</span><span class="o">.</span><span class="n">pass_fasta</span>
<span class="go">&#39;/path/to/current/directory/example1/example1_gRNA_pass.fasta&#39;</span>
</pre></div>
</div>
<p><strong>CLI code (bash)</strong></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --prefix example <span class="m">1</span> <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\ </span><span class="c1">## reference genome</span>
         --indv ref --gene AT1G33560 <span class="se">\ </span><span class="c1">## query genome and gene</span>
         --gc-min <span class="m">0</span> --gc-max <span class="m">1</span> <span class="se">\ </span><span class="c1">## accept all gRNA regardless of GC content</span>
         --auto <span class="c1">## auto-approve minimum gRNA sets</span>
</pre></div>
</div>
<p>Note that the CLI code invokes the full programme, so minimum gRNA set(s) will also be generated. Nevertheless, you can still find a FASTA file of all passing gRNA (that are within the desired feature AND have desired GC content) at ‘/path/to/current/directory/example1/example1_gRNA_pass.fasta’.</p>
<p><strong>CLI code (bash)</strong>: using reference genome aliases</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --prefix example <span class="m">1</span> <span class="se">\</span>
         --reference-set ./arabidopsis_genomes.txt --ref TAIR10 <span class="se">\ </span><span class="c1">## reference genome</span>
         --indv ref --gene AT1G33560 <span class="se">\ </span><span class="c1">## query genome and gene</span>
         --gc-min <span class="m">0</span> --gc-max <span class="m">1</span> <span class="se">\ </span><span class="c1">## accept all gRNA regardless of GC content</span>
         --auto <span class="c1">## auto-approve minimum gRNA sets</span>
</pre></div>
</div>
<p>Here, we use <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code> to give MINORg a mapping of reference genome FASTA and GFF files to more easy to write aliases. In this case, the alias for the combination of ‘./subset_ref_TAIR10.fasta’ and ‘./subset_ref_TAIR10.gff’ is ‘TAIR10’, which we pass to <code class="docutils literal notranslate"><span class="pre">--ref</span></code>.</p>
<p><strong>CLI code (bash)</strong>: using config.ini</p>
<p>Open ./sample_config.ini and replace ‘/path/to/arabidopsis_genomes.txt’ with the absolute path to ./araidopsis_genomes.txt. Save it.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ <span class="nb">export</span> <span class="nv">MINORG_CONFIG</span><span class="o">=</span>./sample_config.ini
$ minorg --prefix example <span class="m">1</span> <span class="se">\</span>
         --indv ref --gene AT1G33560 <span class="se">\ </span><span class="c1">## query genome and gene</span>
         --gc-min <span class="m">0</span> --gc-max <span class="m">1</span> <span class="se">\ </span><span class="c1">## accept all gRNA regardless of GC content</span>
         --auto <span class="c1">## auto-approve minimum gRNA sets</span>
</pre></div>
</div>
<p>If we open ‘./sample_config.ini’, we will find the following lines:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="n">sets</span> <span class="o">=</span> <span class="n">athaliana</span><span class="p">:</span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">athaliana_genomes</span><span class="o">.</span><span class="n">txt</span>
                 <span class="n">arabidopsis</span><span class="p">:</span><span class="o">/&lt;</span><span class="n">absolute</span> <span class="n">path</span><span class="o">&gt;/</span><span class="n">arabidopsis_genomes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>These lines tell MINORg to use ‘arabidopsis’ as an alias for arabidopsis_genomes.txt.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="nb">set</span> <span class="o">=</span> <span class="n">arabidopsis</span>
</pre></div>
</div>
<p>This line tells MINORg to pass arabidopsis_genomes.txt (mapped from alias ‘arabidopsis’) to <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="o">=</span> <span class="n">TAIR10</span>
</pre></div>
</div>
<p>And finally this line tells MINORg to use the FASTA and GFF combination mapped to the alias ‘TAIR10’ in arabidopsis_genomes.txt as the reference genome.</p>
</section>
<section id="example-2">
<h2><span class="section-number">8.2. </span>Example 2<a class="headerlink" href="#example-2" title="Permalink to this headline"></a></h2>
<p>Description of desired outcome</p>
<p>Python
CLI</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="algorithms.html" class="btn btn-neutral float-left" title="7. Algorithms" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="minorg.html" class="btn btn-neutral float-right" title="API documentation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>