<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>4. Tutorial (Command line) &mdash; MINORg 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="5. Tutorial (Python)" href="tutorial_py.html" />
    <link rel="prev" title="3. Configuration" href="configuration.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="modules.html" class="icon icon-home"> MINORg
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">2. Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">3. Configuration</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">4. Tutorial (Command line)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#setting-up-the-tutorial">4.1. Setting up the tutorial</a></li>
<li class="toctree-l2"><a class="reference internal" href="#impt-note-on-executables">4.2. IMPT: Note on executables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#defining-target-sequences">4.3. Defining target sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#user-provided-targets">4.3.1. User-provided targets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reference-gene-s-as-targets">4.3.2. Reference gene(s) as targets</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#single-gene">4.3.2.1. Single gene</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multiple-genes">4.3.2.2. Multiple genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#multiple-and-non-standard-reference">4.3.2.3. Multiple and non-standard reference</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#non-reference-gene-s-as-targets">4.3.3. Non-reference gene(s) as targets</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#annotated-genes">4.3.3.1. Annotated genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#unannotated-genes">4.3.3.2. Unannotated genes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#domain-as-targets">4.3.4. Domain as targets</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#local-database">4.3.4.1. Local database</a></li>
<li class="toctree-l4"><a class="reference internal" href="#remote-database">4.3.4.2. Remote database</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#defining-grna">4.4. Defining gRNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filtering-grna">4.5. Filtering gRNA</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#filter-by-gc-content">4.5.1. Filter by GC content</a></li>
<li class="toctree-l3"><a class="reference internal" href="#filter-by-off-target">4.5.2. Filter by off-target</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#using-total-mismatch-gap-unaligned">4.5.2.1. Using total mismatch/gap/unaligned</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-position-specific-mismatch-gap-unaligned">4.5.2.2. Using position-specific mismatch/gap/unaligned</a></li>
<li class="toctree-l4"><a class="reference internal" href="#pam-less-off-target-check">4.5.2.3. PAM-less off-target check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#skip-off-target-check">4.5.2.4. Skip off-target check</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filter-by-feature">4.5.3. Filter by feature</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#generating-minimum-grna-set-s">4.6. Generating minimum gRNA set(s)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#number-of-sets">4.6.1. Number of sets</a></li>
<li class="toctree-l3"><a class="reference internal" href="#prioritise-non-redundancy">4.6.2. Prioritise non-redundancy</a></li>
<li class="toctree-l3"><a class="reference internal" href="#excluding-grna">4.6.3. Excluding gRNA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#accepting-unknown-checks">4.6.4. Accepting unknown checks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#manually-approve-grna-sets">4.6.5. Manually approve gRNA sets</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#subcommands">4.7. Subcommands</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-seq">4.7.1. Subcommand <code class="docutils literal notranslate"><span class="pre">seq</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-grna">4.7.2. Subcommand <code class="docutils literal notranslate"><span class="pre">grna</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-filter">4.7.3. Subcommand <code class="docutils literal notranslate"><span class="pre">filter</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#gc-check">4.7.3.1. GC check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#background-check">4.7.3.2. Background check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#feature-check">4.7.3.3. Feature check</a></li>
<li class="toctree-l4"><a class="reference internal" href="#combination-of-checks">4.7.3.4. Combination of checks</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#subcommand-minimumset">4.7.4. Subcommand <code class="docutils literal notranslate"><span class="pre">minimumset</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#defining-reference-genomes">4.8. Defining reference genomes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#single-reference-genome">4.8.1. Single reference genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiple-reference-genomes">4.8.2. Multiple reference genomes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#non-standard-reference">4.8.3. Non-standard reference</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#non-standard-genetic-code">4.8.3.1. Non-standard genetic code</a></li>
<li class="toctree-l4"><a class="reference internal" href="#non-standard-gff3-attribute-field-names">4.8.3.2. Non-standard GFF3 attribute field names</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#multithreading">4.9. Multithreading</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_py.html">5. Tutorial (Python)</a></li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">6. Parameters</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">7. Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">8. Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">9. Examples</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="minorg.html">API documentation</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="modules.html">MINORg</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="modules.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">4. </span>Tutorial (Command line)</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorial_cli.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial-command-line">
<h1><span class="section-number">4. </span>Tutorial (Command line)<a class="headerlink" href="#tutorial-command-line" title="Permalink to this headline"></a></h1>
<p>In all the following tutorial, the current directory/working directory is presumed to contain all files in <a class="reference external" href="https://github.com/rlrq/MINORg/tree/master/examples">https://github.com/rlrq/MINORg/tree/master/examples</a>. If you have not downloaded the files, please do so and navigate to the directory that contains them.</p>
<p>Note that all command line code snippets in the following tutorial are for <strong>bash terminal</strong>. You may have to adapt them according to your terminal.</p>
<section id="setting-up-the-tutorial">
<h2><span class="section-number">4.1. </span>Setting up the tutorial<a class="headerlink" href="#setting-up-the-tutorial" title="Permalink to this headline"></a></h2>
<p>To ensure that the examples in this tutorial work, please replace ‘/path/to’ in the files ‘arabidopsis_genomes.txt’, ‘athaliana_genomes.txt’, and ‘subset_genome_mapping.txt’ with the full path to the directory containing the example files.</p>
</section>
<section id="impt-note-on-executables">
<h2><span class="section-number">4.2. </span>IMPT: Note on executables<a class="headerlink" href="#impt-note-on-executables" title="Permalink to this headline"></a></h2>
<p>See: <a class="reference internal" href="parameters.html#blastn-rpsblast-rpsblast-mafft"><span class="std std-ref">blastn, rpsblast/rpsblast+, MAFFT</span></a>, <a class="reference internal" href="parameters.html#bedtools"><span class="std std-ref">BEDTools</span></a></p>
<p>If blastn, rpsblast/rpsblast+, mafft, and/or bedtools is/are not in your command-search path, you will have to append one or more of the appropriate parameters below to your command to tell MINORg where they are.</p>
<p><code class="docutils literal notranslate"><span class="pre">--blastn</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">blastn</span> <span class="pre">executable&gt;</span></code></p>
<ul class="simple">
<li><p>Applicable to: <code class="docutils literal notranslate"><span class="pre">minorg</span></code> (full programme), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">seq</span></code> (subcommand <a class="reference internal" href="#subcommand-seq"><span class="std std-ref">Subcommand seq</span></a>), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">filter</span></code> (<a class="reference internal" href="#subcommand-filter"><span class="std std-ref">Subcommand filter</span></a>)</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">--rpsblast</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">rpsblast</span> <span class="pre">or</span> <span class="pre">rpsblast+</span> <span class="pre">executable&gt;</span></code></p>
<ul class="simple">
<li><p>Applicable to: <code class="docutils literal notranslate"><span class="pre">minorg</span></code> (full programme), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">seq</span></code> (subcommand <a class="reference internal" href="#subcommand-seq"><span class="std std-ref">Subcommand seq</span></a>), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">grna</span></code> (subcommand <a class="reference internal" href="#subcommand-grna"><span class="std std-ref">Subcommand grna</span></a>), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">filter</span></code> (<a class="reference internal" href="#subcommand-filter"><span class="std std-ref">Subcommand filter</span></a>)</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">--mafft</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">MAFFT</span> <span class="pre">executable&gt;</span></code></p>
<ul class="simple">
<li><p>Applicable to: <code class="docutils literal notranslate"><span class="pre">minorg</span></code> (full programme), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">seq</span></code> (subcommand <a class="reference internal" href="#subcommand-seq"><span class="std std-ref">Subcommand seq</span></a>), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">grna</span></code> (subcommand <a class="reference internal" href="#subcommand-grna"><span class="std std-ref">Subcommand grna</span></a>), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">filter</span></code> (<a class="reference internal" href="#subcommand-filter"><span class="std std-ref">Subcommand filter</span></a>)</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">--bedtools</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">directory</span> <span class="pre">containing</span> <span class="pre">BEDTools</span> <span class="pre">executables&gt;</span></code></p>
<ul class="simple">
<li><p>Applicable to: <code class="docutils literal notranslate"><span class="pre">minorg</span></code> (full programme), <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">seq</span></code> (subcommand <a class="reference internal" href="#subcommand-seq"><span class="std std-ref">Subcommand seq</span></a>)</p></li>
</ul>
</section>
<section id="defining-target-sequences">
<h2><span class="section-number">4.3. </span>Defining target sequences<a class="headerlink" href="#defining-target-sequences" title="Permalink to this headline"></a></h2>
<section id="user-provided-targets">
<h3><span class="section-number">4.3.1. </span>User-provided targets<a class="headerlink" href="#user-provided-targets" title="Permalink to this headline"></a></h3>
<p>Let us begin with the simplest MINORg execution:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_100_target --target ./sample_CDS.fasta
Final gRNA sequence<span class="o">(</span>s<span class="o">)</span> have been written to minorg_gRNA_final.fasta
Final gRNA sequence ID<span class="o">(</span>s<span class="o">)</span>, gRNA sequence<span class="o">(</span>s<span class="o">)</span>, and target<span class="o">(</span>s<span class="o">)</span> have been written to minorg_gRNA_final.map

<span class="m">1</span> mutually exclusive gRNA set<span class="o">(</span>s<span class="o">)</span> requested. <span class="m">1</span> set<span class="o">(</span>s<span class="o">)</span> found.
Output files have been generated <span class="k">in</span> /path/to/current/directory/example_100_target
</pre></div>
</div>
<p>The above combination of arguments tells MINORg to generate gRNA from targets in a user-provided FASTA file (<code class="docutils literal notranslate"><span class="pre">--target</span> <span class="pre">./sample_CDS.fasta</span></code>) and to output files into a user-specified directory (<code class="docutils literal notranslate"><span class="pre">--directory</span> <span class="pre">./example_100_target</span></code>). By default, MINORg generates 20 bp gRNA using NGG PAM.</p>
</section>
<section id="reference-gene-s-as-targets">
<h3><span class="section-number">4.3.2. </span>Reference gene(s) as targets<a class="headerlink" href="#reference-gene-s-as-targets" title="Permalink to this headline"></a></h3>
<p>Both examples below specify a reference assembly (<code class="docutils literal notranslate"><span class="pre">--assembly</span> <span class="pre">./subset_ref_TAIR10.fasta</span></code>) and annotation (<code class="docutils literal notranslate"><span class="pre">--annotation</span> <span class="pre">./subset_ref_TAIR10.gff</span></code>) file, allowing MINORg to retrieve the gene sequence as target.</p>
<section id="single-gene">
<h4><span class="section-number">4.3.2.1. </span>Single gene<a class="headerlink" href="#single-gene" title="Permalink to this headline"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_101_singlegene <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff
Extracting reference sequences
Finding homologues
Max acceptable insertion length: <span class="m">15</span>
Final gRNA sequence<span class="o">(</span>s<span class="o">)</span> have been written to minorg_gRNA_final.fasta
Final gRNA sequence ID<span class="o">(</span>s<span class="o">)</span>, gRNA sequence<span class="o">(</span>s<span class="o">)</span>, and target<span class="o">(</span>s<span class="o">)</span> have been written to minorg_gRNA_final.map

<span class="m">1</span> mutually exclusive gRNA set<span class="o">(</span>s<span class="o">)</span> requested. <span class="m">1</span> set<span class="o">(</span>s<span class="o">)</span> found.
Output files have been generated <span class="k">in</span> /path/to/current/directory/example01
</pre></div>
</div>
<p>In the above example, <code class="docutils literal notranslate"><span class="pre">--indv</span> <span class="pre">ref</span></code> tells MINORg to generate gRNA for reference gene(s), and <code class="docutils literal notranslate"><span class="pre">--gene</span> <span class="pre">AT5G45050</span></code> tells MINORg that the target gene is AT5G45050.</p>
</section>
<section id="multiple-genes">
<h4><span class="section-number">4.3.2.2. </span>Multiple genes<a class="headerlink" href="#multiple-genes" title="Permalink to this headline"></a></h4>
<section id="using-gene">
<h5><span class="section-number">4.3.2.2.1. </span>Using <code class="docutils literal notranslate"><span class="pre">--gene</span></code><a class="headerlink" href="#using-gene" title="Permalink to this headline"></a></h5>
<p>See also: <a class="reference internal" href="parameters.html#comma-separated-cli"><span class="std std-ref">Comma-separated (CLI)</span></a>, <a class="reference internal" href="parameters.html#multi-argument-cli"><span class="std std-ref">Multi-argument (CLI)</span></a></p>
<p>To specify multiple genes, simply use <code class="docutils literal notranslate"><span class="pre">--gene</span></code> with comma-separated gene IDs, or <code class="docutils literal notranslate"><span class="pre">--gene</span></code> multiple times</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_102_multigene <span class="se">\</span>
         --indv ref --gene AT5G45050,AT5G45060,AT5G45200,AT5G45210,AT5G45220,AT5G45230,AT5G45240,AT5G45250 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff
</pre></div>
</div>
<p>OR</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_102_multigene <span class="se">\</span>
         --indv ref --gene AT5G45050 --gene AT5G45060 --gene AT5G45200 <span class="se">\</span>
         --gene AT5G45210 --gene AT5G45220 --gene AT5G45230 --gene AT5G45240 --gene AT5G45250 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff
</pre></div>
</div>
</section>
<section id="using-cluster">
<h5><span class="section-number">4.3.2.2.2. </span>Using <code class="docutils literal notranslate"><span class="pre">--cluster</span></code><a class="headerlink" href="#using-cluster" title="Permalink to this headline"></a></h5>
<p>See also: <a class="reference internal" href="configuration.html#id1"><span class="std std-ref">2-level lookup</span></a>, <a class="reference internal" href="parameters.html#comma-separated-cli"><span class="std std-ref">Comma-separated (CLI)</span></a>, <a class="reference internal" href="parameters.html#multi-argument-cli"><span class="std std-ref">Multi-argument (CLI)</span></a></p>
<p>MINORg can also accept preset combinations of genes using <code class="docutils literal notranslate"><span class="pre">--cluster</span></code> and <code class="docutils literal notranslate"><span class="pre">--cluster-set</span></code>. <code class="docutils literal notranslate"><span class="pre">--cluster-set</span></code> accepts a tab-separated lookup file that maps alias(es) to a combinations of genes (see <a class="reference internal" href="configuration.html#cluster"><span class="std std-ref">cluster</span></a> for format). <code class="docutils literal notranslate"><span class="pre">--cluster</span></code> is used to specify the alias of a combination of genes in that lookup file.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_103_cluster <span class="se">\</span>
         --indv ref --cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff
</pre></div>
</div>
<p>The above code snippet is effectively identical to the examples in <a class="reference internal" href="#multiple-genes"><span class="std std-ref">Multiple genes</span></a>.</p>
<p>Like <code class="docutils literal notranslate"><span class="pre">--gene</span></code>, multiple combinations of genes can be specified to <code class="docutils literal notranslate"><span class="pre">--cluster</span></code>. However, unlike <code class="docutils literal notranslate"><span class="pre">--gene</span></code>, each combination will be processed separately (i.e. minimum sets will be separately generated for each combination).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_103_cluster <span class="se">\</span>
         --indv ref --cluster RPS6,TTR1 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff
</pre></div>
</div>
</section>
</section>
<section id="multiple-and-non-standard-reference">
<h4><span class="section-number">4.3.2.3. </span>Multiple and non-standard reference<a class="headerlink" href="#multiple-and-non-standard-reference" title="Permalink to this headline"></a></h4>
<p>See: <a class="reference internal" href="#defining-reference-genomes"><span class="std std-ref">Defining reference genomes</span></a></p>
<p>Multiple reference genomes may be useful when generating gRNA across species boundaries. See <a class="reference internal" href="#multiple-reference-genomes"><span class="std std-ref">Multiple reference genomes</span></a> for how to specify and use multiple reference genomes.</p>
<p>Some reference genomes may require non-standard genetic code (applicable only with the use of <code class="docutils literal notranslate"><span class="pre">--domain</span></code>) or have unusual attribute field names in their GFF3 annotation files. See <a class="reference internal" href="#non-standard-genetic-code"><span class="std std-ref">Non-standard genetic code</span></a> for how to specify non-standard genetic codes and <a class="reference internal" href="#non-standard-gff3-attribute-field-names"><span class="std std-ref">Non-standard GFF3 attribute field names</span></a> for how to specify mapping of unusual GFF3 attribute field names to standard field names.</p>
</section>
</section>
<section id="non-reference-gene-s-as-targets">
<h3><span class="section-number">4.3.3. </span>Non-reference gene(s) as targets<a class="headerlink" href="#non-reference-gene-s-as-targets" title="Permalink to this headline"></a></h3>
<section id="annotated-genes">
<h4><span class="section-number">4.3.3.1. </span>Annotated genes<a class="headerlink" href="#annotated-genes" title="Permalink to this headline"></a></h4>
<p>If your target genes have been annotated in their non-reference genomes (i.e. you have a GFF3 file containing annotations of your targets), you can use <a class="reference internal" href="#reference-gene-s-as-targets"><span class="std std-ref">Reference gene(s) as targets</span></a> if you have a single non-reference genome, or <a class="reference internal" href="#multiple-reference-genomes"><span class="std std-ref">Multiple reference genomes</span></a> if you have multiple non-reference genomes. In either case, you may treat your non-reference genome the same way you would a reference genome.</p>
</section>
<section id="unannotated-genes">
<h4><span class="section-number">4.3.3.2. </span>Unannotated genes<a class="headerlink" href="#unannotated-genes" title="Permalink to this headline"></a></h4>
<section id="using-extend-gene-and-extend-cds">
<h5><span class="section-number">4.3.3.2.1. </span>Using <code class="docutils literal notranslate"><span class="pre">--extend-gene</span></code> and <code class="docutils literal notranslate"><span class="pre">--extend-cds</span></code><a class="headerlink" href="#using-extend-gene-and-extend-cds" title="Permalink to this headline"></a></h5>
<p>See also: <a class="reference internal" href="parameters.html#extended-genome"><span class="std std-ref">Extended genome</span></a></p>
<p>If you have both genomic and CDS-only sequences of your target genes but not a GFF3 annotation file, MINORg can infer coding regions (CDS) for your target genes using <code class="docutils literal notranslate"><span class="pre">--extend-gene</span></code> and <code class="docutils literal notranslate"><span class="pre">--extend-cds</span></code>. See <a class="reference internal" href="parameters.html#extended-genome"><span class="std std-ref">Extended genome</span></a> for how to name your sequences to ensure proper mapping of CDS to genes.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_104_ext <span class="se">\</span>
         --indv ref --gene AT1G10920 <span class="se">\</span>
         --extend-gene ./sample_gene.fasta --extend-cds ./sample_CDS.fasta
</pre></div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">--extend-gene</span></code> and <code class="docutils literal notranslate"><span class="pre">--extend-cds</span></code> effectively add new genes to the reference genome, so they can be used just like any reference gene. Therefore, they can also be used in combination with <code class="docutils literal notranslate"><span class="pre">--query</span></code> or <code class="docutils literal notranslate"><span class="pre">--indv</span></code>.</p>
</section>
<section id="using-query">
<h5><span class="section-number">4.3.3.2.2. </span>Using <code class="docutils literal notranslate"><span class="pre">--query</span></code><a class="headerlink" href="#using-query" title="Permalink to this headline"></a></h5>
<p>See also: <a class="reference internal" href="algorithms.html#non-reference-homologue-inference"><span class="std std-ref">Non-reference homologue inference</span></a>, <a class="reference internal" href="parameters.html#multi-argument-cli"><span class="std std-ref">Multi-argument (CLI)</span></a></p>
<p>If you would like MINORg to infer homologues genes in non-reference genomes, you can use <code class="docutils literal notranslate"><span class="pre">--query</span></code> to specify the FASTA files of those non-reference genomes. You may provide multiple non-reference genomes by using <code class="docutils literal notranslate"><span class="pre">--query</span></code> multiple times.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_105_query <span class="se">\</span>
         --query ./subset_9654.fasta --query ./subset_9655.fasta <span class="se">\</span>
         --gene AT1G10920 <span class="se">\</span>
         --extend-gene ./sample_gene.fasta --extend-cds ./sample_CDS.fasta
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--query</span></code> can be used in combination with <code class="docutils literal notranslate"><span class="pre">--indv</span></code>. For inference parameters, see <a class="reference internal" href="algorithms.html#non-reference-homologue-inference"><span class="std std-ref">Non-reference homologue inference</span></a>.</p>
</section>
<section id="using-indv">
<h5><span class="section-number">4.3.3.2.3. </span>Using <code class="docutils literal notranslate"><span class="pre">--indv</span></code><a class="headerlink" href="#using-indv" title="Permalink to this headline"></a></h5>
<p>See also: <a class="reference internal" href="algorithms.html#non-reference-homologue-inference"><span class="std std-ref">Non-reference homologue inference</span></a>, <a class="reference internal" href="configuration.html#id1"><span class="std std-ref">2-level lookup</span></a>, <a class="reference internal" href="parameters.html#comma-separated-cli"><span class="std std-ref">Comma-separated (CLI)</span></a>, <a class="reference internal" href="parameters.html#multi-argument-cli"><span class="std std-ref">Multi-argument (CLI)</span></a></p>
<p>You can also use <code class="docutils literal notranslate"><span class="pre">--indv</span></code> to ask MINORg to infer homologues genes in non-reference genomes. Similar to <code class="docutils literal notranslate"><span class="pre">--clusters</span></code>, MINORg accepts a lookup file for non-reference genomes using <code class="docutils literal notranslate"><span class="pre">--genome-set</span></code> (see <a class="reference internal" href="configuration.html#genome"><span class="std std-ref">genome</span></a> for format) and one or more non-reference genome alias using <code class="docutils literal notranslate"><span class="pre">--indv</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_106_indv <span class="se">\</span>
         --indv <span class="m">9654</span>,9655 --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --gene AT1G10920 <span class="se">\</span>
         --extend-gene ./sample_gene.fasta --extend-cds ./sample_CDS.fasta
</pre></div>
</div>
<p>The above code snippet is effectively identical to the example in <a class="reference internal" href="#using-query"><span class="std std-ref">Using --query</span></a>.</p>
<p><code class="docutils literal notranslate"><span class="pre">--indv</span></code> can be used in combination with <code class="docutils literal notranslate"><span class="pre">--query</span></code>. For inference parameters, see <a class="reference internal" href="algorithms.html#non-reference-homologue-inference"><span class="std std-ref">Non-reference homologue inference</span></a>.</p>
</section>
</section>
</section>
<section id="domain-as-targets">
<h3><span class="section-number">4.3.4. </span>Domain as targets<a class="headerlink" href="#domain-as-targets" title="Permalink to this headline"></a></h3>
<p>MINORg allows users to specify the identifier of an RPS-BLAST position-specific scoring matrix (PSSM-Id) to further restrict the target sequence to a given domain associated with the PSSM-Id. This could be particularly useful when designing gRNA for genes that do not share conserved domain structures but do share a domain that you wish to knock out. <code class="docutils literal notranslate"><span class="pre">--domain</span></code> can also be used with <code class="docutils literal notranslate"><span class="pre">--query</span></code> or <code class="docutils literal notranslate"><span class="pre">--indv</span></code>.</p>
<section id="local-database">
<h4><span class="section-number">4.3.4.1. </span>Local database<a class="headerlink" href="#local-database" title="Permalink to this headline"></a></h4>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_107_domain <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --rpsblast /path/to/rpsblast/executable --db /path/to/rpsblast/db <span class="se">\</span>
         --domain <span class="m">214815</span>
</pre></div>
</div>
<p>In the above example, gRNA will be generated for the WRKY domain (PSSM-Id 214815 as of CDD database v3.18) of the gene AT5G45050. Users are responsible for providing the PSSM-Id of a domain that exists in the gene. Unlike other examples, the database (<code class="docutils literal notranslate"><span class="pre">--db</span></code>) is not provided as part of the example files. You will have to download it yourself. See <a class="reference internal" href="parameters.html#rps-blast-local-database"><span class="std std-ref">RPS-BLAST local database</span></a> for more information.</p>
</section>
<section id="remote-database">
<h4><span class="section-number">4.3.4.2. </span>Remote database<a class="headerlink" href="#remote-database" title="Permalink to this headline"></a></h4>
<p>While it is in theory possible to use the remote CDD database &amp; servers instead of local ones, the <code class="docutils literal notranslate"><span class="pre">--remote</span></code> option for the ‘rpsblast’/’rpsblast+’ command from the BLAST+ package has never worked for me. In any case, if your version of local rpsblast is able to access the remote database, you can use <code class="docutils literal notranslate"><span class="pre">--remote-rps</span></code> instead of <code class="docutils literal notranslate"><span class="pre">--db</span> <span class="pre">/path/to/rpsblast/db</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_107_domain <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --rpsblast /path/to/rpsblast/executable --remote-rps <span class="se">\</span>
         --domain <span class="m">214815</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="defining-grna">
<h2><span class="section-number">4.4. </span>Defining gRNA<a class="headerlink" href="#defining-grna" title="Permalink to this headline"></a></h2>
<p>See also: <a class="reference internal" href="parameters.html#pam"><span class="std std-ref">PAM</span></a></p>
<p>By default, MINORg generates 20 bp gRNA using SpCas9’s NGG PAM. You may specify other gRNA length using <code class="docutils literal notranslate"><span class="pre">--length</span></code> and other PAM using <code class="docutils literal notranslate"><span class="pre">--pam</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_108_grna <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --length <span class="m">19</span> --pam Cas12a
</pre></div>
</div>
<p>In the example above, MINORg will generate 19 bp gRNA (<code class="docutils literal notranslate"><span class="pre">--length</span> <span class="pre">19</span></code>) using Cas12a’s unusual 5’ PAM pattern (TTTV&lt;gRNA&gt;) (<code class="docutils literal notranslate"><span class="pre">--pam</span> <span class="pre">Cas12a</span></code>). MINORg has several built-in PAMs (see <a class="reference internal" href="parameters.html#preset-pam-patterns"><span class="std std-ref">Preset PAM patterns</span></a> for options), and also supports customisable PAM patterns using ambiguous bases and regular expressions (see <a class="reference internal" href="parameters.html#pam"><span class="std std-ref">PAM</span></a> for format).</p>
</section>
<section id="filtering-grna">
<h2><span class="section-number">4.5. </span>Filtering gRNA<a class="headerlink" href="#filtering-grna" title="Permalink to this headline"></a></h2>
<p>MINORg supports 3 different gRNA filtering options, all of which can be used together.</p>
<section id="filter-by-gc-content">
<h3><span class="section-number">4.5.1. </span>Filter by GC content<a class="headerlink" href="#filter-by-gc-content" title="Permalink to this headline"></a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_109_gc <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --gc-min <span class="m">0</span>.2 --gc-max <span class="m">0</span>.8
</pre></div>
</div>
<p>In the above example, MINORg will exclude gRNA with less than 20% (<code class="docutils literal notranslate"><span class="pre">--gc-min</span> <span class="pre">0.2</span></code>) or greater than 80% (<code class="docutils literal notranslate"><span class="pre">--gc-max</span> <span class="pre">0.8</span></code>) GC content. By default, minimum GC content is 30% and maximum is 70%.</p>
</section>
<section id="filter-by-off-target">
<h3><span class="section-number">4.5.2. </span>Filter by off-target<a class="headerlink" href="#filter-by-off-target" title="Permalink to this headline"></a></h3>
<p>See: <a class="reference internal" href="algorithms.html#off-target-assessment"><span class="std std-ref">Off-target assessment</span></a></p>
<section id="using-total-mismatch-gap-unaligned">
<h4><span class="section-number">4.5.2.1. </span>Using total mismatch/gap/unaligned<a class="headerlink" href="#using-total-mismatch-gap-unaligned" title="Permalink to this headline"></a></h4>
<p>See: <a class="reference internal" href="algorithms.html#total-mismatch-gap-unaligned"><span class="std std-ref">Total mismatch/gap/unaligned</span></a></p>
<p>Thresholds for total number of mismatches or gaps (and unaligned positions) required for an off-target gRNA hit to be considered non-problematic are controlled by <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code> and <code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code> respectively. See <a class="reference internal" href="algorithms.html#total-mismatch-gap-unaligned"><span class="std std-ref">Total mismatch/gap/unaligned</span></a> for more.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_110_ot_ref <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --screen-reference <span class="se">\</span>
         --background ./subset_ref_Araly2.fasta --background ./subset_ref_Araha1.fasta <span class="se">\</span>
         --ot-indv <span class="m">9654</span>,9655 --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --ot-gap <span class="m">2</span> --ot-mismatch <span class="m">2</span>
</pre></div>
</div>
<p>In the above example, MINORg will screen gRNA for off-targets in:</p>
<ul class="simple">
<li><p>The reference genome (<code class="docutils literal notranslate"><span class="pre">--screen-reference</span></code>)</p></li>
<li><p>Two different FASTA files (<code class="docutils literal notranslate"><span class="pre">--background</span> <span class="pre">./subset_Araly2.fasta</span> <span class="pre">--background</span> <span class="pre">./subset_Araha1.fasta</span></code>)</p></li>
<li><p>Two non-reference genomes (<code class="docutils literal notranslate"><span class="pre">--ot-indv</span> <span class="pre">9654,9655</span> <span class="pre">--genome-set</span> <span class="pre">./subset_genome_mapping.txt</span></code>)</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">--ot-indv</span></code> functions similarly to <code class="docutils literal notranslate"><span class="pre">--indv</span></code> in that it requires <code class="docutils literal notranslate"><span class="pre">--genome-set</span></code>, except that <code class="docutils literal notranslate"><span class="pre">--ot-indv</span></code> specifies non-refernece genomes for off-target assessment</p></li>
<li><p>Note that any AT5G45050 homologues in these two genomes will NOT be masked. This means that only gRNA that do not target any AT5G45050 homologues in these two non-reference genomes will pass this off-target check.
* To mask homologues in these genomes, you will need to provide a FASTA file containing the sequences of their homologues using <code class="docutils literal notranslate"><span class="pre">--mask</span> <span class="pre">&lt;FASTA&gt;</span></code>. You may use subcommand <code class="docutils literal notranslate"><span class="pre">seq</span></code> (see <a class="reference internal" href="#subcommand-seq"><span class="std std-ref">Subcommand seq</span></a>) to identify these homologues.</p></li>
</ul>
</li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code> and <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code> control the minimum number of gaps or mismatches off-target gRNA hits must have to be considered non-problematic; any gRNA with at least one problematic gRNA hit will be excluded. See <a class="reference internal" href="algorithms.html#off-target-assessment"><span class="std std-ref">Off-target assessment</span></a> for more on the off-target assessment algorithm.</p>
<p>In the case above, <code class="docutils literal notranslate"><span class="pre">--screen-reference</span></code> is actually redundant as the genome from which targets are obtained (which, because of <code class="docutils literal notranslate"><span class="pre">--indv</span> <span class="pre">ref</span></code>, is the reference genome) are automatically included for background check. However, in the example below, when the targets are from <strong>non-reference genomes</strong>, the reference genome is not automatically included for off-target assessment and thus <code class="docutils literal notranslate"><span class="pre">--screen-reference</span></code> is NOT redundant. Additionally, do note that the genes passed to <code class="docutils literal notranslate"><span class="pre">--gene</span></code> are masked in the reference genome, such that any gRNA hits to them are NOT considered off-target and will NOT be excluded.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_111_ot_nonref <span class="se">\</span>
         --indv <span class="m">9654</span> --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --screen-ref --background ./subset_ref_Araly2.fasta --background ./subset_ref_Araha1.fasta <span class="se">\</span>
         --ot-indv <span class="m">9655</span> <span class="se">\</span>
         --ot-gap <span class="m">2</span> --ot-mismatch <span class="m">2</span>
</pre></div>
</div>
</section>
<section id="using-position-specific-mismatch-gap-unaligned">
<h4><span class="section-number">4.5.2.2. </span>Using position-specific mismatch/gap/unaligned<a class="headerlink" href="#using-position-specific-mismatch-gap-unaligned" title="Permalink to this headline"></a></h4>
<p>See: <a class="reference internal" href="algorithms.html#position-specific-mismatch-gap-unaligned"><span class="std std-ref">Position-specific mismatch/gap/unaligned</span></a></p>
<p>Finer control of off-target definition can be achieved using <a class="reference internal" href="minorg.MINORg.html#id6" title="minorg.MINORg.MINORg.ot_pattern"><code class="xref py py-attr docutils literal notranslate"><span class="pre">ot_pattern</span></code></a>, which allows users to provide a pattern that specifies different thresholds for different positions along a gRNA. Unlike <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code> and <code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code>, which specify the <strong>LOWER-bound of NON-problematic</strong> hits, <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span></code> specifies <strong>UPPER-bound of PROBLEMATIC</strong> hits. By default, unaligned positions will be treated as mismatches, but this behaviour can be altered by raising <code class="docutils literal notranslate"><span class="pre">--ot-unaligned-as-mismatch-unset</span></code>. See <a class="reference internal" href="parameters.html#off-target-pattern"><span class="std std-ref">Off-target pattern</span></a> for how to build an off-target pattern, and <a class="reference internal" href="algorithms.html#position-specific-mismatch-gap-unaligned"><span class="std std-ref">Position-specific mismatch/gap/unaligned</span></a> for more on how unaligned positions can be counted.</p>
<p>When <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span></code> is specified, <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code> and <code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code> will be ignored.</p>
<p>The following example is identical to the first in <a class="reference internal" href="tutorial_py.html#using-total-mismatch-gap-unaligned"><span class="std std-ref">Using total mismatch/gap/unaligned</span></a>, except <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code> and <code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code> are replaced with <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_112_ot_ref_pattern <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --screen-reference <span class="se">\</span>
         --background ./subset_ref_Araly2.fasta --background ./subset_ref_Araha1.fasta <span class="se">\</span>
         --ot-indv <span class="m">9654</span>,9655 --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --ot-pattern <span class="s1">&#39;2mg-5-,0mg4&#39;</span>
</pre></div>
</div>
<p>In the above example, <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span> <span class="pre">'0mg-4,2mg-5-'</span></code> means that MINORg will discard any gRNA with at least one off-target hit where:</p>
<ul class="simple">
<li><p>There are no mismatches or gaps between positions -4 and -1, and there are no more than 2 mismatches or gaps from position -5 to the 5’ end.</p></li>
</ul>
</section>
<section id="pam-less-off-target-check">
<h4><span class="section-number">4.5.2.3. </span>PAM-less off-target check<a class="headerlink" href="#pam-less-off-target-check" title="Permalink to this headline"></a></h4>
<p>By default, MINORg does NOT check for the presence of PAM sites next to potential off-target hits. You may override this behaviour using <code class="docutils literal notranslate"><span class="pre">--ot-pam</span></code>. This tells MINORg to mark off-target hits that fail the <code class="docutils literal notranslate"><span class="pre">--ot-gap</span></code> or <code class="docutils literal notranslate"><span class="pre">--ot-mismatch</span></code> thresholds (or match <code class="docutils literal notranslate"><span class="pre">--ot-pattern</span></code>) as problematic ONLY IF there is a PAM site nearby.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_113_ot_pamless <span class="se">\</span>
         --indv <span class="m">9654</span> --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --screen-ref --background ./subset_ref_Araly2.fasta --background ./subset_ref_Araha1.fasta <span class="se">\</span>
         --ot-indv <span class="m">9655</span> <span class="se">\</span>
         --ot-gap <span class="m">2</span> --ot-mismatch <span class="m">2</span> <span class="se">\</span>
         --ot-pam
</pre></div>
</div>
</section>
<section id="skip-off-target-check">
<h4><span class="section-number">4.5.2.4. </span>Skip off-target check<a class="headerlink" href="#skip-off-target-check" title="Permalink to this headline"></a></h4>
<p>To skip off-target check entirely, use <code class="docutils literal notranslate"><span class="pre">--skip-bg-check</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_114_skipbgcheck <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --skip-bg-check
</pre></div>
</div>
</section>
</section>
<section id="filter-by-feature">
<h3><span class="section-number">4.5.3. </span>Filter by feature<a class="headerlink" href="#filter-by-feature" title="Permalink to this headline"></a></h3>
<p>See: <a class="reference internal" href="algorithms.html#within-feature-inference"><span class="std std-ref">Within-feature inference</span></a></p>
<p>By default, when <code class="docutils literal notranslate"><span class="pre">--gene</span></code> is used, MINORg restricts gRNA to coding regions (CDS). For more on how MINORg does this for inferred, unannotated homologues, see <a class="reference internal" href="algorithms.html#within-feature-inference"><span class="std std-ref">Within-feature inference</span></a>. You may change the feature type in which to design gRNA using <code class="docutils literal notranslate"><span class="pre">--feature</span></code>. See column 3 of your GFF3 file for valid feature types (see <a class="reference external" href="https://en.wikipedia.org/wiki/General_feature_format">https://en.wikipedia.org/wiki/General_feature_format</a> for more on GFF file format).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_115_withinfeature <span class="se">\</span>
         --indv ref --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --feature three_prime_UTR
</pre></div>
</div>
</section>
</section>
<section id="generating-minimum-grna-set-s">
<h2><span class="section-number">4.6. </span>Generating minimum gRNA set(s)<a class="headerlink" href="#generating-minimum-grna-set-s" title="Permalink to this headline"></a></h2>
<section id="number-of-sets">
<h3><span class="section-number">4.6.1. </span>Number of sets<a class="headerlink" href="#number-of-sets" title="Permalink to this headline"></a></h3>
<p>By default, MINORg outputs a single gRNA set covering all targets. You may request more (mutually exclusive) sets using <code class="docutils literal notranslate"><span class="pre">--set</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_116_set <span class="se">\</span>
         --indv ref --cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --set <span class="m">5</span>
</pre></div>
</div>
</section>
<section id="prioritise-non-redundancy">
<h3><span class="section-number">4.6.2. </span>Prioritise non-redundancy<a class="headerlink" href="#prioritise-non-redundancy" title="Permalink to this headline"></a></h3>
<p>By default, MINORg selects gRNA for sets using these criteria in decreasing order of priority:</p>
<ol class="arabic simple">
<li><p>Coverage (of as yet uncovered targets)</p></li>
<li><p>Proximity to 5’ end</p></li>
<li><p>Non-redundancy</p></li>
</ol>
<p>Proximity is only assessed when there is a tie for coverage, and non-redundancy when there is a tie for both coverage and proximity. You may flip the priority of proximity and non-redundancy using <code class="docutils literal notranslate"><span class="pre">--prioritise-nr</span></code> if you prefer to minimise multiple edits in a single target when using a single set of gRNA. (The priority of coverage is NOT modifiable.)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_117_nr <span class="se">\</span>
         --indv ref --cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --prioritise-nr
</pre></div>
</div>
</section>
<section id="excluding-grna">
<h3><span class="section-number">4.6.3. </span>Excluding gRNA<a class="headerlink" href="#excluding-grna" title="Permalink to this headline"></a></h3>
<p>You may specify gRNA sequences to exclude from any final gRNA set using <code class="docutils literal notranslate"><span class="pre">--exclude</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_118_exclude <span class="se">\</span>
         --indv ref --cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --exclude ./sample_exclude_RPS6.fasta
</pre></div>
</div>
<p>The gRNA names in the file passed to <code class="docutils literal notranslate"><span class="pre">--exclude</span></code> do not matter. Only the sequences are used when determining whether to exclude a gRNA.</p>
</section>
<section id="accepting-unknown-checks">
<h3><span class="section-number">4.6.4. </span>Accepting unknown checks<a class="headerlink" href="#accepting-unknown-checks" title="Permalink to this headline"></a></h3>
<p>Sometimes, not all filtering checks (GC, background, and feature) are set for all sequences. This is not an issue if you use the full programme (i.e. <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">&lt;arguments&gt;</span></code>), but may be relevant if you are re-generating sets using the ‘minimumset’ subcommand (i.e. <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">minimumset</span> <span class="pre">&lt;arguments&gt;</span></code>) with a modified mapping file OR a mapping file from the ‘filter’ subcommand where not all filters have been applied.</p>
<p>Let us take a look at ‘sample_custom_check.map’, where we’ve added a custom check called ‘my_custom_check’ in the last column:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gRNA</span> <span class="nb">id</span>       <span class="n">gRNA</span> <span class="n">sequence</span>   <span class="n">target</span> <span class="nb">id</span>       <span class="n">target</span> <span class="n">sense</span>    <span class="n">gRNA</span> <span class="n">strand</span>     <span class="n">start</span>   <span class="n">end</span>     <span class="n">group</span>   <span class="n">background</span>      <span class="n">GC</span>      <span class="n">feature</span> <span class="n">my_custom_check</span>
<span class="n">gRNA_001</span>      <span class="n">CTTCATCTTCTTCTCGAAAT</span>    <span class="n">targetA</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">8</span>       <span class="mi">27</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_001</span>      <span class="n">CTTCATCTTCTTCTCGAAAT</span>    <span class="n">targetB</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">80</span>      <span class="mi">99</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetA</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">37</span>      <span class="mi">56</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">NA</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetB</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">286</span>     <span class="mi">305</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetC</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">109</span>     <span class="mi">128</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">fail</span>
<span class="n">gRNA_002</span>      <span class="n">GATGTTTTCTTGAGCTTCAG</span>    <span class="n">targetD</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">110</span>     <span class="mi">129</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">fail</span>
<span class="n">gRNA_003</span>      <span class="n">ATGTTTTCTTGAGCTTCAGA</span>    <span class="n">targetB</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">38</span>      <span class="mi">57</span>      <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="n">NA</span>
<span class="n">gRNA_003</span>      <span class="n">ATGTTTTCTTGAGCTTCAGA</span>    <span class="n">targetC</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">287</span>     <span class="mi">306</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
<span class="n">gRNA_003</span>      <span class="n">ATGTTTTCTTGAGCTTCAGA</span>    <span class="n">targetD</span> <span class="n">NA</span>      <span class="o">+</span>       <span class="mi">110</span>     <span class="mi">129</span>     <span class="mi">1</span>       <span class="k">pass</span>    <span class="k">pass</span>    <span class="n">NA</span>      <span class="k">pass</span>
</pre></div>
</div>
<p>There are three possible values for check status: ‘pass’, ‘fail’, and ‘NA’.</p>
<p>An invalid/unset check is an ‘NA’. If a check is unset for all entries (as is the case with the check ‘feature’ here), it will be ignored (i.e. the check is treated as ‘pass’ for all entries). However, when a check has been set for some entries but not others (as is the case with the ‘my_custom_check’ check here), MINORg will treat invalid/unset checks as ‘fail’ by default. This is because there isn’t enough information on whether this constitutes a pass or fail for the check, and MINORg prefers to be conservative when outputting gRNA. You may override this behaviour using the <code class="docutils literal notranslate"><span class="pre">--accept-invalid</span></code>. By doing so, MINORg will treat ‘NA’ as ‘pass’ for all checks.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg minimumset --directory ./example_119_acceptinvalid <span class="se">\</span>
                    --map ./sample_custom_check.map <span class="se">\</span>
                    --accept-invalid
</pre></div>
</div>
</section>
<section id="manually-approve-grna-sets">
<h3><span class="section-number">4.6.5. </span>Manually approve gRNA sets<a class="headerlink" href="#manually-approve-grna-sets" title="Permalink to this headline"></a></h3>
<p>You may opt to manually inspect each gRNA set before MINORg write them to file using the <code class="docutils literal notranslate"><span class="pre">--manual</span></code> flag.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_120_manual --target ./sample_CDS.fasta
         --manual
     ID      sequence <span class="o">(</span>Set <span class="m">1</span><span class="o">)</span>
     gRNA_001        GGAATACAAGAGATTATCGA
Hit <span class="s1">&#39;x&#39;</span> to <span class="k">continue</span> <span class="k">if</span> you are satisfied with these sequences. Otherwise, enter the sequence ID or sequence of an undesirable gRNA <span class="o">(</span><span class="k">case</span>-sensitive<span class="o">)</span> and hit the <span class="k">return</span> key to update this list: x
Final gRNA sequence<span class="o">(</span>s<span class="o">)</span> have been written to minorg_gRNA_final.fasta
Final gRNA sequence ID<span class="o">(</span>s<span class="o">)</span>, gRNA sequence<span class="o">(</span>s<span class="o">)</span>, and target<span class="o">(</span>s<span class="o">)</span> have been written to minorg_gRNA_final.map

<span class="m">1</span> mutually exclusive gRNA set<span class="o">(</span>s<span class="o">)</span> requested. <span class="m">1</span> set<span class="o">(</span>s<span class="o">)</span> found.
Output files have been generated <span class="k">in</span> /path/to/current/directory/example_119_manual
</pre></div>
</div>
</section>
</section>
<section id="subcommands">
<h2><span class="section-number">4.7. </span>Subcommands<a class="headerlink" href="#subcommands" title="Permalink to this headline"></a></h2>
<p>MINORg comprises of four main steps:</p>
<ol class="arabic simple">
<li><p>Target sequence identification</p></li>
<li><p>Candidate gRNA generation</p></li>
<li><p>gRNA filtering</p></li>
<li><p>Minimum gRNA set generation</p></li>
</ol>
<p>As users may only wish to execute a subset of these steps instead of the full programme, MINORg also provides four subcommands corresponding to these four steps:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">seq</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">grna</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">filter</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">minimumset</span></code></p></li>
</ol>
<p>The subcommands may be useful if you already have a preferred off-target/on-target assessment software. In this case, you may execute subcommands <code class="docutils literal notranslate"><span class="pre">seq</span></code> and <code class="docutils literal notranslate"><span class="pre">grna</span></code>, submit the gRNA output by MINORg for off-target/on-target assessment, update the .map file output by MINORg with the status of each gRNA for that off-target/on-target assessment, and execute <code class="docutils literal notranslate"><span class="pre">minimumset</span></code> to obtain a desired number of minimum gRNA sets.</p>
<section id="subcommand-seq">
<h3><span class="section-number">4.7.1. </span>Subcommand <code class="docutils literal notranslate"><span class="pre">seq</span></code><a class="headerlink" href="#subcommand-seq" title="Permalink to this headline"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">seq</span></code> subcommand identifies target sequences, whether by extracting them from a reference genome or inferring homologues in unannotated genomes. All parameters described in <a class="reference internal" href="#defining-target-sequences"><span class="std std-ref">Defining target sequences</span></a> (except <code class="docutils literal notranslate"><span class="pre">--target</span></code>) and <a class="reference internal" href="#defining-reference-genomes"><span class="std std-ref">Defining reference genomes</span></a> apply.</p>
<p>This step will output target sequences into a file ending with ‘_targets.fasta’.</p>
<p>To use this subcommand, simply replace the command <code class="docutils literal notranslate"><span class="pre">minorg</span></code> with <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">seq</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg seq --directory ./example_121_subcmdseq <span class="se">\</span>
             --query ./subset_9654.fasta --query ./subset_9655.fasta <span class="se">\</span>
             --gene AT1G10920 <span class="se">\</span>
             --extend-gene ./sample_gene.fasta --extend-cds ./sample_CDS.fasta
</pre></div>
</div>
</section>
<section id="subcommand-grna">
<h3><span class="section-number">4.7.2. </span>Subcommand <code class="docutils literal notranslate"><span class="pre">grna</span></code><a class="headerlink" href="#subcommand-grna" title="Permalink to this headline"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">grna</span></code> subcommand generates gRNA within target sequences. It incorporates parts of the <code class="docutils literal notranslate"><span class="pre">seq</span></code> and <code class="docutils literal notranslate"><span class="pre">filter</span></code> subcommands in order to provide rudimentary filtering for gRNA within specific GFF3 features (e.g. CDS) for reference genes as well as by GC content. All parameters described in <a class="reference internal" href="#defining-target-sequences"><span class="std std-ref">Defining target sequences</span></a> (except those related to homology discovery in unannotated genomes such as <code class="docutils literal notranslate"><span class="pre">query</span></code>, <code class="docutils literal notranslate"><span class="pre">indv</span></code>, and <code class="docutils literal notranslate"><span class="pre">genome-set</span></code>), <a class="reference internal" href="#defining-reference-genomes"><span class="std std-ref">Defining reference genomes</span></a>, <a class="reference internal" href="#defining-grna"><span class="std std-ref">Defining gRNA</span></a>, and <a class="reference internal" href="#filter-by-gc-content"><span class="std std-ref">Filter by GC content</span></a>, and <a class="reference internal" href="#filter-by-feature"><span class="std std-ref">Filter by feature</span></a> apply.</p>
<p>Unlike the full programme or the <code class="docutils literal notranslate"><span class="pre">seq</span></code> subcommand, however, <code class="docutils literal notranslate"><span class="pre">--indv</span> <span class="pre">ref</span></code> is not necessary to specify reference genes as target. As this subcommand does not support homologue discovery, if <code class="docutils literal notranslate"><span class="pre">--gene</span></code> or <code class="docutils literal notranslate"><span class="pre">--cluster</span></code> is used, <code class="docutils literal notranslate"><span class="pre">--indv</span> <span class="pre">ref</span></code> will automatically be filled since unannotated genomes are not allowed.</p>
<p>This step will output target sequences into a file ending with ‘_targets.fasta’ if <code class="docutils literal notranslate"><span class="pre">--target</span></code> was not used. gRNA sequences will be written into files ending with ‘_gRNA_all.fasta’ (for all candidate gRNA) and ‘_gRNA_pass.fasta’ (for candidate gRNA that pass GC and feature checks). A file ending with ‘_gRNA_all.map’ that maps gRNA to their targets will also be generated. You may optionally specify the location of the FASTA and .map output files using:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--out-fasta</span></code>: path to output file that originally ends with ‘_gRNA_all.fasta’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--out-pass</span></code>: path to output file that originally ends with ‘_gRNA_pass.fasta’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--out-map</span></code>: path to output file that originally ends with ‘_gRNA_all.map’</p></li>
</ul>
<p>To use this subcommand, simply replace the command <code class="docutils literal notranslate"><span class="pre">minorg</span></code> with <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">grna</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg grna --directory ./example_122_subcmdgrna <span class="se">\</span>
              --cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
              --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
              --length <span class="m">19</span> --pam Cas12a <span class="se">\</span>
              --feature three_prime_UTR <span class="se">\</span>
              --gc-min <span class="m">0</span>.2 --gc-max <span class="m">0</span>.8 <span class="se">\</span>
              --out-map ./example_120.map
</pre></div>
</div>
</section>
<section id="subcommand-filter">
<h3><span class="section-number">4.7.3. </span>Subcommand <code class="docutils literal notranslate"><span class="pre">filter</span></code><a class="headerlink" href="#subcommand-filter" title="Permalink to this headline"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">filter</span></code> subcommand takes in a compulsory MINORg .map file (<code class="docutils literal notranslate"><span class="pre">--map</span></code>) and rewrites some/all checks. You should specify the checks you wish to re-assess using some combination of <code class="docutils literal notranslate"><span class="pre">--gc-check</span></code>, <code class="docutils literal notranslate"><span class="pre">--background-check</span></code>, and/or <code class="docutils literal notranslate"><span class="pre">--feature-check</span></code> flags OR <code class="docutils literal notranslate"><span class="pre">--check-all</span></code> to raise all three flags. For in-place modification of the .map file, use <code class="docutils literal notranslate"><span class="pre">--in-place</span></code>. Otherwise, MINORg will write a new file using the default naming format of ‘&lt;prefix&gt;_gRNA_all.map’ (this may still overwrite the original file if the directory and prefix are identical to what was used to generate the original file).</p>
<p>gRNA sequences will be written into files ending with ‘_gRNA_all.fasta’ (for all candidate gRNA) and ‘_gRNA_pass.fasta’ (for candidate gRNA that pass the updated checks). A file ending with ‘_gRNA_all.map’ that maps gRNA to their targets will also be generated with the updated check statuses. As with subcommand <code class="docutils literal notranslate"><span class="pre">grna</span></code>, you may optionally specify the location of the FASTA and .map output files using:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--out-fasta</span></code>: path to output file that originally ends with ‘_gRNA_all.fasta’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--out-pass</span></code>: path to output file that originally ends with ‘_gRNA_pass.fasta’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--out-map</span></code>: path to output file that originally ends with ‘_gRNA_all.map’</p></li>
</ul>
<p>To use this subcommand, simply replace the command <code class="docutils literal notranslate"><span class="pre">minorg</span></code> with <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">filter</span></code>.</p>
<p>In all cases, you may rename the gRNA using <code class="docutils literal notranslate"><span class="pre">--rename</span> <span class="pre">&lt;FASTA&gt;</span></code>, where the FASTA file contains the gRNA sequences you wish to rename with sequence IDs of the names you wish to rename them to.</p>
<section id="gc-check">
<h4><span class="section-number">4.7.3.1. </span>GC check<a class="headerlink" href="#gc-check" title="Permalink to this headline"></a></h4>
<p>All parameters described in <a class="reference internal" href="#filter-by-gc-content"><span class="std std-ref">Filter by GC content</span></a> apply.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg filter --directory ./example_123_subcmdfilter_gc <span class="se">\</span>
                --map ./sample_custom_check.map <span class="se">\</span>
                --gc-check --gc-min <span class="m">0</span>.2 --gc-max <span class="m">0</span>.8
</pre></div>
</div>
</section>
<section id="background-check">
<h4><span class="section-number">4.7.3.2. </span>Background check<a class="headerlink" href="#background-check" title="Permalink to this headline"></a></h4>
<p>All parameters described in <a class="reference internal" href="#filter-by-off-target"><span class="std std-ref">Filter by off-target</span></a> apply. Additionally, you should supply target sequences using <code class="docutils literal notranslate"><span class="pre">--target</span></code> so that MINORg can mask them (this tells MINORg that any gRNA hits to them is in fact on-target and NOT off-target). Any additional sequences to be masked may be provided using <code class="docutils literal notranslate"><span class="pre">--mask</span> <span class="pre">&lt;FASTA&gt;</span></code>. If you are using <code class="docutils literal notranslate"><span class="pre">--screen-ref</span></code> to include reference genome(s) (see <a class="reference internal" href="#multiple-reference-genomes"><span class="std std-ref">Multiple reference genomes</span></a> for how to specify multiple reference genomes) in the off-target screen, you may specify reference genes to be masked using <code class="docutils literal notranslate"><span class="pre">--mask-gene</span></code> or <code class="docutils literal notranslate"><span class="pre">--mask-cluster</span></code> (unlike <code class="docutils literal notranslate"><span class="pre">--cluster</span></code>, all clusters passed to <code class="docutils literal notranslate"><span class="pre">--mask-cluster</span></code> will be processed simultaneously; i.e. there will not be separate executions for each cluster).</p>
<p>Let us first generate a .map file for filtering.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_124_subcmdfilter_bg_pt1 <span class="se">\</span>
         --indv <span class="m">9654</span>,9655 --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
         --skip-bg-check
</pre></div>
</div>
<p>In the code above, we skipped off-target check by raising the <code class="docutils literal notranslate"><span class="pre">--skip-bg-check</span></code> flag. But we’ve changed out mind and would like to screen the reference genome and the non-reference genomes that these targets are from AND we don’t want our gRNA to be able to target any genes in ‘subset_9944.fasta’ and ‘subset_9947’. We can do that using the <code class="docutils literal notranslate"><span class="pre">filter</span></code> subcommand.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg filter --directory ./example_124_subcmdfilter_bg_pt2 <span class="se">\</span>
                --map ./example_123_subcmdfilter_bg_pt1/minorg_RPS6/minorg_RPS6_gRNA_all.map <span class="se">\</span>
                --background-check <span class="se">\</span>
                --target ./example_123_subcmdfilter_bg_pt1/minorg_RPS6/minorg_RPS6_gene_targets.fasta <span class="se">\</span>
                --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
                --screen-ref <span class="se">\</span>
                --mask-cluster RPS6 --cluster-set ./subset_cluster_mapping.txt <span class="se">\</span>
                --ot-indv <span class="m">9654</span>,9655,9944,9947 --genome-set ./subset_genome_mapping.txt
</pre></div>
</div>
<p>The above code may be a little unwieldy. However, if the target identification step of MINORg takes a while to run (for example when the genome files are large and take forever to process), you may prefer not to re-run the full MINORg programme with updated parameters and instead use the <code class="docutils literal notranslate"><span class="pre">filter</span></code> subcommand on files that have already been generated. You should then use the <code class="docutils literal notranslate"><span class="pre">minimumset</span></code> subcommand (see <a class="reference internal" href="#subcommand-minimumset"><span class="std std-ref">Subcommand minimumset</span></a>) to regenerate minimum sets using your update .map file.</p>
</section>
<section id="feature-check">
<h4><span class="section-number">4.7.3.3. </span>Feature check<a class="headerlink" href="#feature-check" title="Permalink to this headline"></a></h4>
<p>All parameters described in <a class="reference internal" href="#filter-by-feature"><span class="std std-ref">Filter by feature</span></a> apply. Additionally, you will need to provide a FASTA file of target sequences (using <code class="docutils literal notranslate"><span class="pre">--target</span> <span class="pre">&lt;FASTA&gt;</span></code>), reference genome(s) (see <a class="reference internal" href="#defining-reference-genomes"><span class="std std-ref">Defining reference genomes</span></a>), and genes (using <code class="docutils literal notranslate"><span class="pre">--gene</span> <span class="pre">&lt;gene</span> <span class="pre">IDs&gt;</span></code> or <code class="docutils literal notranslate"><span class="pre">--cluster</span> <span class="pre">&lt;cluster</span> <span class="pre">alias&gt;</span></code>). The specified reference gene(s) will be extracted from the reference genome(s) and aligned with target sequence(s) in order for MINORg to infer feature boundaries in target sequence(s). See <a class="reference internal" href="algorithms.html#within-feature-inference"><span class="std std-ref">Within-feature inference</span></a> for the algorithm of how feature boundaries are inferred.</p>
<p>Do note that unlike the full programme or the <code class="docutils literal notranslate"><span class="pre">seq</span></code> subcommand, all clusters passed to <code class="docutils literal notranslate"><span class="pre">--cluster</span></code> will be processed simultaneously (i.e. there will not be separate executions for each cluster).</p>
<p>Let us first generate a .map file for filtering.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_125_subcmdfilter_feature_pt1 <span class="se">\</span>
         --indv <span class="m">9654</span>,9655 --genome-set ./subset_genome_mapping.txt <span class="se">\</span>
         --gene AT5G45050 <span class="se">\</span>
         --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff
</pre></div>
</div>
<p>By default, MINORg sets the desired feature to ‘CDS’. You can re-assess and overwrite the ‘feature’ check in the .map file to only allow gRNA in the 3’ UTR using <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">filter</span></code> with the <code class="docutils literal notranslate"><span class="pre">--feature-check</span></code> flag raised.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg filter --directory ./example_125_subcmdfilter_feature_pt2 <span class="se">\</span>
                --map ./example_124_subcmdfilter_feature_pt1/minorg/minorg_gRNA_all.map <span class="se">\</span>
                --feature-check <span class="se">\</span>
                --target ./example_124_subcmdfilter_feature_pt1/minorg/minorg_gene_targets.fasta <span class="se">\</span>
                --gene AT5G45050 <span class="se">\</span>
                --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
                --feature three_prime_UTR
</pre></div>
</div>
</section>
<section id="combination-of-checks">
<h4><span class="section-number">4.7.3.4. </span>Combination of checks<a class="headerlink" href="#combination-of-checks" title="Permalink to this headline"></a></h4>
<p>You can execute all checks (or some combination of them) in a single <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">filter</span></code> command as well, if you wish. Just make sure that you raise the appropriate flag(s).</p>
<p>To use some combination of checks, simply raise the relevant flags (<code class="docutils literal notranslate"><span class="pre">--gc-check</span></code>, <code class="docutils literal notranslate"><span class="pre">--background-check</span></code>, and/or <code class="docutils literal notranslate"><span class="pre">--feature-check</span></code>). In the example below, we filter the gRNA generated by full MINORg execution in <a class="reference internal" href="#feature-check"><span class="std std-ref">Feature check</span></a> by both GC content (<code class="docutils literal notranslate"><span class="pre">--gc-check</span></code>) as well as gene feature (<code class="docutils literal notranslate"><span class="pre">--feature-check</span></code>).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg filter --directory ./example_126_subcmdfilter_gcfeature <span class="se">\</span>
                --map ./example_124_subcmdfilter_feature_pt1/minorg/minorg_gRNA_all.map <span class="se">\</span>
                --gc-check <span class="se">\</span>
                --gc-min <span class="m">0</span>.2 --gc-max <span class="m">0</span>.8 <span class="se">\</span>
                --feature-check <span class="se">\</span>
                --target ./example_124_subcmdfilter_feature_pt1/minorg/minorg_gene_targets.fasta <span class="se">\</span>
                --gene AT5G45050 <span class="se">\</span>
                --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
                --feature three_prime_UTR
</pre></div>
</div>
<p>To execute all checks, use <code class="docutils literal notranslate"><span class="pre">--check-all</span></code>. In the example below, we filter the gRNA generated by full MINORg execution in <a class="reference internal" href="#feature-check"><span class="std std-ref">Feature check</span></a> by all checks.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg filter --directory ./example_127_subcmdfilter_all <span class="se">\</span>
                --map ./example_124_subcmdfilter_feature_pt1/minorg/minorg_gRNA_all.map <span class="se">\</span>
                --check-all <span class="se">\</span>
                --gc-min <span class="m">0</span>.2 --gc-max <span class="m">0</span>.8 <span class="se">\ </span><span class="c1">## GC</span>
                --target ./example_124_subcmdfilter_feature_pt1/minorg/minorg_gene_targets.fasta <span class="se">\ </span><span class="c1">## feature</span>
                --gene AT5G45050 <span class="se">\</span>
                --assembly ./subset_ref_TAIR10.fasta --annotation ./subset_ref_TAIR10.gff <span class="se">\</span>
                --feature three_prime_UTR <span class="se">\</span>
                --screen-ref --mask-gene AT5G45050 <span class="c1">## off-target</span>
</pre></div>
</div>
</section>
</section>
<section id="subcommand-minimumset">
<h3><span class="section-number">4.7.4. </span>Subcommand <code class="docutils literal notranslate"><span class="pre">minimumset</span></code><a class="headerlink" href="#subcommand-minimumset" title="Permalink to this headline"></a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">minimumset</span></code> subcommand generates mutually exclusive minimum set(s) of gRNA, where each set is capable of covering all targets. All parameters described in <a class="reference internal" href="#generating-minimum-grna-set-s"><span class="std std-ref">Generating minimum gRNA set(s)</span></a> apply.</p>
<p>This step will write final gRNA sequences into a file ending with ‘_gRNA_final.fasta’. A file ending with ‘_gRNA_final.map’ that maps gRNA to their targets will also be generated. You may optionally specify the location of the FASTA and .map output files using:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--out-fasta</span></code>: path to output file that originally ends with ‘_gRNA_final.fasta’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--out-map</span></code>: path to output file that originally ends with ‘_gRNA_final.map’</p></li>
</ul>
<p><strong>NOTE:</strong> Unlike subcommands <code class="docutils literal notranslate"><span class="pre">grna</span></code> and <code class="docutils literal notranslate"><span class="pre">filter</span></code>, <code class="docutils literal notranslate"><span class="pre">--out-fasta</span></code> and <code class="docutils literal notranslate"><span class="pre">--out-map</span></code> are used to specify output files for <strong>FINAL</strong> gRNA sets, not all candidate gRNA.</p>
<p>To use this subcommand, simply replace the command <code class="docutils literal notranslate"><span class="pre">minorg</span></code> with <code class="docutils literal notranslate"><span class="pre">minorg</span> <span class="pre">grna</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg minimumset --directory ./example_128_subcmdminimumset
                    --map ./example_105_query/minorg/minorg_gRNA_all.map <span class="se">\</span>
                    --target ./example_105_query/minorg/minorg_gene_targets.fasta <span class="se">\</span>
                    --set <span class="m">5</span> --manual --prioritise-nr
</pre></div>
</div>
<p>In order for MINORg to better assess a gRNA’s proximity to the 5’ end (of hopefully sense strand) of a target in the event a tie-breaker is necessary, it is strongly suggested that target sequences be provided using <code class="docutils literal notranslate"><span class="pre">--target</span> <span class="pre">&lt;FASTA&gt;</span></code> so MINORg knows how long a target sequence is. This is especially so if the target sequences are antisense ones (you can check this using the .map file) generated by MINORg’s inferences of homologues in unannotated genomes.</p>
</section>
</section>
<section id="defining-reference-genomes">
<h2><span class="section-number">4.8. </span>Defining reference genomes<a class="headerlink" href="#defining-reference-genomes" title="Permalink to this headline"></a></h2>
<section id="single-reference-genome">
<h3><span class="section-number">4.8.1. </span>Single reference genome<a class="headerlink" href="#single-reference-genome" title="Permalink to this headline"></a></h3>
<p>See examples in <a class="reference internal" href="#reference-gene-s-as-targets"><span class="std std-ref">Reference gene(s) as targets</span></a>.</p>
</section>
<section id="multiple-reference-genomes">
<h3><span class="section-number">4.8.2. </span>Multiple reference genomes<a class="headerlink" href="#multiple-reference-genomes" title="Permalink to this headline"></a></h3>
<p>See also: <a class="reference internal" href="parameters.html#reference"><span class="std std-ref">Reference</span></a>, <a class="reference internal" href="configuration.html#id1"><span class="std std-ref">2-level lookup</span></a>, <a class="reference internal" href="parameters.html#comma-separated-cli"><span class="std std-ref">Comma-separated (CLI)</span></a>, <a class="reference internal" href="parameters.html#multi-argument-cli"><span class="std std-ref">Multi-argument (CLI)</span></a></p>
<p>Similar to <code class="docutils literal notranslate"><span class="pre">--clusters</span></code> and <code class="docutils literal notranslate"><span class="pre">--indv</span></code>, MINORg accepts a lookup file for reference genomes using <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code> and one or more reference genome alias using <code class="docutils literal notranslate"><span class="pre">--reference</span></code>. See <a class="reference internal" href="parameters.html#reference"><span class="std std-ref">Reference</span></a> for a more comprehensive overview and <a class="reference internal" href="configuration.html#reference"><span class="std std-ref">reference</span></a> for lookup file format.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_129_multiref <span class="se">\</span>
         --indv ref --gene AT1G33560,AL1G47950.v2.1,Araha.3012s0003.v1.1 <span class="se">\</span>
         --reference tair10,araly2,araha1 --reference-set ./arabidopsis_genomes.txt
</pre></div>
</div>
<p>In the example above, MINORg will design gRNA for 3 highly conserved paralogues in 3 different species. Note that you should be careful that any gene IDs you use should either be unique across all reference genomes OR be shared only among your target genes. Otherwise, MINORg will treat any undesired genes with the same gene IDs as targets as well.</p>
<p>Non-standard genetic codes and mapping of non-standard attribute field names for multiple genomes should be specified in the lookup file passed to <code class="docutils literal notranslate"><span class="pre">--reference-set</span></code>. See <a class="reference internal" href="configuration.html#reference"><span class="std std-ref">reference</span></a> for file format.</p>
</section>
<section id="non-standard-reference">
<h3><span class="section-number">4.8.3. </span>Non-standard reference<a class="headerlink" href="#non-standard-reference" title="Permalink to this headline"></a></h3>
<section id="non-standard-genetic-code">
<h4><span class="section-number">4.8.3.1. </span>Non-standard genetic code<a class="headerlink" href="#non-standard-genetic-code" title="Permalink to this headline"></a></h4>
<p>When using <code class="docutils literal notranslate"><span class="pre">--domain</span></code>, users should ensure that the correct genetic code is specified, as MINORg has to first translate CDS into peptides for domain search using RPS-BLAST. The default genetic code is the Standard Code. Please refer to <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a> for genetic code numbers and names.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorgpy --directory ./example_129_geneticcode <span class="se">\</span>
           --indv ref --gene gene-Q0275 <span class="se">\</span>
           --assembly ./subset_ref_yeast_mt.fasta --annotation ./subset_ref_yeast_mt.gff <span class="se">\</span>
           --domain <span class="m">366140</span> --genetic-code <span class="m">3</span>
</pre></div>
</div>
<p>In the above example, the gene ‘gene-Q0275’ is a yeast mitochondrial gene, and <code class="docutils literal notranslate"><span class="pre">--domain</span> <span class="pre">366140</span></code> specifies the PSSM-Id for the COX3 domain in the Cdd v3.18 RPS-BLAST database. The genetic code number for yeast mitochondrial code is ‘3’.</p>
<p>As a failsafe, MINORg does not terminate translated peptide sequences at the first stop codon. This ensures that any codons after an incorrectly translated premature stop codon will still be translated. Typically, a handful of mistranslated codons can still result in the correct RPS-BLAST domain hits, although hit scores may be slightly lower. Nevertheless, to ensure maximum accuracy, the correct genetic code is preferred.</p>
</section>
<section id="non-standard-gff3-attribute-field-names">
<h4><span class="section-number">4.8.3.2. </span>Non-standard GFF3 attribute field names<a class="headerlink" href="#non-standard-gff3-attribute-field-names" title="Permalink to this headline"></a></h4>
<p>See also: <a class="reference internal" href="parameters.html#attribute-modification"><span class="std std-ref">Attribute modification</span></a></p>
<p>MINORg requires standard attribute field names in GFF3 files in order to properly map subfeatures to their parent features (e.g. map CDS to mRNA, and mRNA to gene). Non-standard field names should be mapped to standard ones using <code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code> (for ‘attribute modification’).</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorgpy --directory ./example_130_attrmod <span class="se">\</span>
           --indv ref --gene Os01t0100100 <span class="se">\</span>
           --assembly ./subset_ref_irgsp.fasta --annotation ./subset_ref_irgsp.gff <span class="se">\</span>
           --attr-mod <span class="s1">&#39;mRNA:Parent=Locus_id&#39;</span>
</pre></div>
</div>
<p>The IRGSP 1.0 reference genome for rice (<em>Oryza sativa</em> subsp. Nipponbare) uses a non-standard attribute field name for mRNA entries in their GFF3 file. Instead of ‘Parent’, which is the standard name of the field used to map a feature to its parent feature, mRNA entries in the IRGSP 1.0 annotation use ‘Locus_id’. See <a class="reference internal" href="parameters.html#attribute-modification"><span class="std std-ref">Attribute modification</span></a> for more details on how to format the input to <code class="docutils literal notranslate"><span class="pre">--attr-mod</span></code>.</p>
</section>
</section>
</section>
<section id="multithreading">
<h2><span class="section-number">4.9. </span>Multithreading<a class="headerlink" href="#multithreading" title="Permalink to this headline"></a></h2>
<p>MINORg supports multi-threading in order to process files in parallel. Any excess threads may also be used for BLAST. This is most useful when you are querying multiple genomes (whether using <code class="docutils literal notranslate"><span class="pre">--query</span></code> or <code class="docutils literal notranslate"><span class="pre">--indv</span></code>), have multiple reference genomes (<code class="docutils literal notranslate"><span class="pre">--reference</span></code>), or multiple background sequences (<code class="docutils literal notranslate"><span class="pre">--background</span></code>).</p>
<p>To run MINORg with parallel processing, use <code class="docutils literal notranslate"><span class="pre">--thread</span> <span class="pre">&lt;number</span> <span class="pre">of</span> <span class="pre">threads&gt;</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ minorg --directory ./example_131_thread <span class="se">\</span>
         --query ./subset_9654.fasta --query ./subset_9655.fasta <span class="se">\</span>
         --gene AT1G10920 <span class="se">\</span>
         --extend-gene ./sample_gene.fasta --extend-cds ./sample_CDS.fasta <span class="se">\</span>
         --thread <span class="m">2</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="configuration.html" class="btn btn-neutral float-left" title="3. Configuration" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial_py.html" class="btn btn-neutral float-right" title="5. Tutorial (Python)" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Rachelle R.Q. Lee.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>